File: Estruturas_de_Dados/abstratos/Pilha/pilha.py
# Como você pode ver aqui, e esta sera uma verdade pra todas
# as próxmas estrutras, de certa forma, a gente usa a Lista Encadeada
# pra construir a Pilha, modificando suas funções para de fato
# se comportar como uma Pilha

class Stack_Node:

    def __init__(self, data):
        self.data = data
        self.next = None
    
    def __repr__(self):
        return str(self.data)

class Stack():
    def __init__(self, maximum_size = 255, top_node:Stack_Node = None):

        self.top = top_node
        self.maximum_size = maximum_size
        self.number_elements = 0
                
    def show_stack(self):
        
        print("\nPilha: [", end="")
        
        temp_node = self.top

        while temp_node:

            print(temp_node, end = ", ")
            temp_node = temp_node.next

        print("\b\b]\n")
        
        return
        
    def add_stack(self, value):
        
        # Esse código em especifico usa um límite de elementos, mas tirando isso,
        # É exatamente igual a função de quase mesmo nome da Lista Encadeada

        if self.maximum_size > self.number_elements:        
            
            new_node = Stack_Node(value)
            new_node.next = self.top
            self.top = new_node
            self.number_elements += 1

            print(f"Node de valor {new_node.data} adicionado a pilha")

            return
            
        else:
            
            print("\nA pilha está cheia")

            return

    def get_top(self):

        return self.top    
    
    def pop_stack(self):

        # Essa operação é igualmente inversa a função de adicionar itens,
        # e apesar de ser meio confuso o uso de "next", pensa em next aqui
        # como pegando próxmio item abaixo, e não necessáriamente a esquerda ou direita
        
        if self.number_elements > 0:

            removed_node = self.top
            self.top = self.top.next

            print(f"Node de valor {removed_node.data} retirado da pilha")

            return removed_node.data
        
        else:
            print("\nA pilha está vázia")


pilha = Stack(12)

pilha.add_stack(1)
pilha.add_stack(5)
pilha.add_stack(327)

pilha.show_stack()

pilha.pop_stack()

pilha.show_stack()

pilha.add_stack(27)

pilha.show_stack()

pilha.pop_stack()
pilha.pop_stack()

pilha.show_stack()

pilha.add_stack(84)
pilha.add_stack(55)
pilha.add_stack(18)
pilha.add_stack(97)

pilha.show_stack()

pilha.pop_stack()
pilha.pop_stack()

pilha.show_stack()

File: Estruturas_de_Dados/abstratos/Pilha/Pilha.java
package Complexo.Pilha;

public class Pilha {
    
    public static void main(String[] args) {
        
        Stack pilha = new Stack();

        pilha.AddStack(1);
        pilha.AddStack(5);
        pilha.AddStack(327);

        pilha.ShowStack();

        pilha.PopStack();

        pilha.ShowStack();

        pilha.AddStack(27);

        pilha.ShowStack();

        pilha.PopStack();
        pilha.PopStack();

        pilha.ShowStack();

        pilha.AddStack(84);
        pilha.AddStack(55);
        pilha.AddStack(18);
        pilha.AddStack(97);

        pilha.ShowStack();

        pilha.PopStack();
        pilha.PopStack();

        pilha.ShowStack();


    }

}

class StackNode {

    int data;
    StackNode next;

    StackNode(int data, StackNode next) {

        this.data = data;
        this.next = next;

    }

    StackNode(int data) {
        this(data, null);
    }

}

class Stack {

    int maximumSize = 255;
    int numberElements = 0;
    StackNode topNode;

    Stack(int maximumSize, StackNode topNode) {
        this.maximumSize = maximumSize;
        this.topNode = topNode;
    }

    Stack(int maximumSize) {
        this(maximumSize, null);
    }

    Stack() {
        this(255, null);
    }

    void AddStack(int value) {
        if (this.maximumSize > this.numberElements) {
            StackNode newNode = new StackNode(value, this.topNode);
            this.topNode = newNode; 
            this.numberElements++; 

            System.out.println("O node " + topNode.data + " foi adicionado a pilha");

        } else {
            System.out.println("O stack ta cheio!");
        }
    }

    Integer PopStack() {

        if (this.numberElements > 0) {

            StackNode removedNode = this.topNode;
            this.topNode = this.topNode.next;

            System.out.println("O node " + removedNode.data + " foi removido da pilha");

            return removedNode.data;

        }

        else System.out.println("O stack esta vazio!");
        
        return null;

    }

    void ShowStack() {

        if (this.topNode != null) {
            
            StackNode temporary_node = this.topNode;
            System.out.print("[");

            while (temporary_node != null) { 

                System.out.print(temporary_node.data);

                if (temporary_node.next != null) {
                    System.out.print(", ");
                }

                temporary_node = temporary_node.next;
                
            }

            System.out.print("]");
            System.err.println("");

        } else {
            System.out.println("O stack esta vazio");
        }
    }
}